ingestion/src/kldata.rs

```diff
@@ -262,116 +276,131 @@ pub async fn filter_and_label_kldata<'a>(
-
-        let transaction = conn.transaction().await?;
-
-        let (sensor, lvl) = in_datum
-            .id
-            .sensor_and_level
-            .map(|both| (Some(both.0), Some(both.1)))
-            .unwrap_or((None, None));
-
-        let obsinn_label_result = transaction
-            .query_opt(
-                &query_get_obsinn,
-                &[
-                    &chunk.station_id,
-                    &chunk.type_id,
-                    &in_datum.id.param_code,
-                    &lvl,
-                    &sensor,
-                ],
-            )
-            .await?;
-
-        let timeseries_id: i32 = match obsinn_label_result {
-            Some(row) => row.get(0),
-            None => {
-                // create new timeseries
-                // TODO: currently we create a timeseries with null location
-                // In the future the location column should be moved to the timeseries metadata table
-                let timeseries_id = transaction
-                    .query_one(
-                        "INSERT INTO public.timeseries (fromtime) VALUES ($1) RETURNING id",
-                        &[&in_datum.timestamp],
-                    )
-                    .await?
-                    .get(0);
+    let mut out_chunks = Vec::with_capacity(chunks.len());
+    for chunk in chunks {
+        let mut data = Vec::with_capacity(chunk.observations.len());
+
+        for in_datum in chunk.observations {
+            // get the conversion first, so we avoid wasting a tsid if it doesn't exist
+            let param = param_conversions
+                .get(&in_datum.id.param_code)
+                .ok_or_else(|| {
+                    Error::Parse(format!(
+                        "unrecognised param_code '{}'",
+                        in_datum.id.param_code
+                    ))
+                })?;
+
+            // TODO: With some changes to this function, we could potentially move its call outside
+            // the loop body. For one thing, the station permit checks, if done in a separate
+            // function, would apply to all observations. Since we know the param-specific permits
+            // are barely used, we could also pre-emptively check all param permits outside the
+            // loop.
+            if !timeseries_is_open(
+                permit_table.clone(),
+                chunk.station_id,
+                chunk.type_id,
+                param.id,
+            )? {
+                // TODO: log that the timeseries is closed? Mostly useful for tests
+                #[cfg(feature = "integration_tests")]
+                eprintln!("station {}: timeseries is closed", chunk.station_id);
+                continue;
+            }
 
-                // create obsinn label
-                transaction
-                    .execute(
-                        "INSERT INTO labels.obsinn \
+            let transaction = conn.transaction().await?;
+
+            let (sensor, lvl) = in_datum
```

Comment:
Any particular reason for using _lvl_ instead of _level_? The latter is more readable.

Reply:
I believe we used `lvl` to match what it's called in obsinn/kvalobs. No objection from me if you feel strongly about changing it, I agree `level` is more readable.

Reply:
No it's fine, it's not that unreadable I guess. And it's a good point that it should match obsinn/kvalobs. But see my other comment about dependency to obsinn in general.

Reply:
Alright, I guess we leave it unchanged then.